Add resizeable backend VNC support

matt335672 opened the pull request:
This has been discussed in the past - see #259 and #287.

The VNC backend still seems to be with us, and still seems to have a few useful use cases.  At our place, developers working at home are are using Citrix to connect in to a Windows farm and from there are using XRDP to connect to the Linux workstations. Some of them have reported noticeably better performance using Xvnc rather than Xorg in this configuration.

This PR supports the VNC resizing and multi-screen features used by tigervnc.  I've tested both resizing and multi-screen on CentOS 7.5 (tigervnc 1.8). and resizing on xubuntu 18.04 (tigervnc 1.7).  I can't currently test multi-monitor on xubuntu due to a lack of kit.

I've made the feature optional to avoid breaking workflows used by existing users. So by default there's no change in behaviour. Users wishing to use the new feature can set "code=1" in xrdp.ini. This value is used to inform sesman that the VNC session is resizeable, and also to tell the VNC backend to request the ExtendedDesktopSize pseudo-encoding from Xvnc.

The module uses resize-on-reconnect in the same way that the Xorg backend does. When a single screen is used, the use can resize sessions on-the-fly by using 'xrandr -s', or by using the appropriate desktop utility.

For review purposes, I've split the PR into three commits. One is simply running astyle on the vnc module. Another is adding the extra functionality to the backend (plus removing some magic numbers), and the last is the changes to SCP and sesman. The last of these is the bit I'm least confident about. At the moment I haven't completely separated resizable VNC sessions from non-resizeable VNC sessions in sesman itself.  Would that be a better idea?

metalefty commented:
Actually, I was working on adding resizing support. Mine is still WIP. I review your implementation and combine with mine. Thank you very much for your continuous contributions.

matt335672 commented:
That sounds like a very good idea - I'm sure you understand the implications for the scp library and sesman better than I do.
I'll leave the PR open for now in case you've got any queries on my implementation, or you'd like me to make other changes.

metalefty commented:
Started working on neutrinolabs extended-vnc branch based on your work:
https://github.com/neutrinolabs/xrdp/tree/extended-vnc

metalefty commented:
@matt335672 
I'm thinking of removing SCP_SESSION_TYPE_RESIZEABLE_VNC to simplify.  I don't want to care about code=0/1. It should be automatic. If VNC server supports resizing (ExtendedDesktopSize), resize the existing VNC session. If resizing is not supported, connect to the existing session with current size. I'd appreciate if you let me know concerns.

This will change curren behaviour but most people using VNC session don't want to create new session when reconnect from different size.

matt335672 commented:
Well, when I proposed the PR, I didn't think you'd be happy with changing existing behaviour. 

If you are happy with that, I very much agree that it's worth removing SCP_SESSION_TYPE_RESIZEABLE_VNC. It makes the code simpler, and all the backends should now behave the same way. You may even be able to remove the 'D' session allocation policy from the session manager as the session manager won't need to store sizes any more - all the size information can be managed by the vnc module.

I suspect you're right in your assumption that most people would prefer to reconnect to an existing session. There may be people out there using the old behaviour, but I can't think that XRDP is adding much value for them.

bolkedebruin commented:
@metalefty can this be reopened again and reconsidered? We now need to arrange apply this patch out of bounds due to the speed difference between VNC and for the exact reasons @matt335672 stated.

metalefty commented:
Hello, actually @matt335672 did most work (thanks!). Then I make some changes to his work. However, there's no progress because of the lack of my time. So the bottleneck is certainly me. I'm really sorry about that. I should leave it to others rather than damming up the development at my place.

@matt335672 can you continue further development as I mentioned at https://github.com/neutrinolabs/xrdp/pull/1343#issuecomment-523757292 ? Sorry about I made no progress.

matt335672 commented:
Hi both,

Yes, I'm happy to do some further work on this.

It'll take me a few days to get to it. When I raised this PR, I was working as a sysadmin to a bunch of software developers with twin monitor setups. At the moment I'm at home without even a twin monitor setup, so I'm going to have to work out how I can set something up for testing and the inevitable bug-fixes that will be needed after we get something out.

@bolkedebruin - if you tell me which version of XRDP you're working with at the moment I'll rebase my extended-vnc branch on [https://github.com/matt335672/xrdp](https://github.com/matt335672/xrdp) to that as soon as I can. If you want to pull from there it might help your existing workflows, and if you want to change versions we can rebase again. If you're happy to test and comment on the changes I'll be making before release that could be quite useful for both of us I think.

bolkedebruin commented:
Awesome! We applied your patch to 0.9.10 but would like it to apply to 0.9.12. We haven't tried yet in this case. We have been running the patch for over 6 months without any issues AFAIK btw. So I would argue patch clean up would be enough.

matt335672 commented:
Hi @bolkedebruin 

I've rebased to v0.9.12 and the branch builds cleanly. For the reasons stated above I haven't tested this yet, but you're welcome to work with it.

I think that was what you wanted - please correct me if not.

Met vriendelijke groet

bolkedebruin commented:
Thanks! Honestly what I would like is to have this in the main branch. I know your patch is working well ;-)

Dank

bolkedebruin commented:
@metalefty anything left for you to be able to merge this?

metalefty commented:
@bolkedebruin Let me some time to test.

metalefty commented on sesman/tools/sesrun.c:
```diff
@@ -61,10 +65,14 @@ main(int argc, char **argv)
     if (argc == 1)
     {
-        g_printf("xrdp session starter v0.1\n");
+        g_printf("xrdp session starter v" PACKAGE_VERSION "\n");
```
Good point.

matt335672 commented:
Now I'm a bit confused.

@metalefty - do you want me to work on removing SCP_SESSION_TYPE_RESIZEABLE_VNC? Happy to do that or leave it as you suggest.

metalefty commented:
@matt335672 

Yes, I'm thinking of like this.
* Remove `SCP_SESSION_TYPE_RESIZEABLE_VNC`
* Always allow VNC to connect existing session (like xorgxrdp session)
* Detect if VNC supports resizing
    * If VNC supports resizing, resize the session to client size
    * If VNC doesn't support resizing, resize the client to server size

matt335672 commented:
I've just added a commit to this branch which implements the changes requested by @metalefty 
- Removes SCP_SESSION_TYPE_RESIZEABLE_VNC
- By default VNC connects to existing session
- Detect if VNC supports resizing at connect time
  - If VNC supports resizing, resize the session to client size
  - If VNC doesn't support resizing, resize the client to server size

I've also fixed a number of bugs, the most important of which are:-
- Windows monitor layouts where the primary monitor was not the left-top monitor now work.
- The Windows client can be resized to single monitor layout when configured as multi-monitor but connecting to a single screen VNC server.

The resize detection has been moved to a separate state machine which runs after the connect which I think makes the logic clearer. I've also added a variable to allow the ExtendedDesktop pseudo-encoding support to be disabled if a serious problem is found with either my changes or a particuat VNC server.

I've tried to use doxygen-compatible function headers where required. 

So far I've tested on tigervnc and an x11vnc session on the console. I'll try a few other VNC servers.

Any comments or testing are very welcome.

Nexarian commented:
So this is for VNC only, right?

What about compatibility with this:
https://github.com/neutrinolabs/xrdp/issues/448?

On Thu, May 21, 2020 at 6:57 AM matt335672 <notifications@github.com> wrote:

> I've just added a commit to this branch which implements the changes
> requested by @metalefty <https://github.com/metalefty>
>
>    - Removes SCP_SESSION_TYPE_RESIZEABLE_VNC
>    - By default VNC connects to existing session
>    - Detect if VNC supports resizing at connect time
>       - If VNC supports resizing, resize the session to client size
>       - If VNC doesn't support resizing, resize the client to server size
>
> I've also fixed a number of bugs, the most important of which are:-
>
>    - Windows monitor layouts where the primary monitor was not the
>    left-top monitor now work.
>    - The Windows client can be resized to single monitor layout when
>    configured as multi-monitor but connecting to a single screen VNC server.
>
> The resize detection has been moved to a separate state machine which runs
> after the connect which I think makes the logic clearer. I've also added a
> variable to allow the ExtendedDesktop pseudo-encoding support to be
> disabled if a serious problem is found with either my changes or a
> particuat VNC server.
>
> I've tried to use doxygen-compatible function headers where required.
>
> So far I've tested on tigervnc and an x11vnc session on the console. I'll
> try a few other VNC servers.
>
> Any comments or testing are very welcome.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/neutrinolabs/xrdp/pull/1343#issuecomment-632021784>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UBHPIFEGKA25KI3TJWVLRSUCIFANCNFSM4HJ5BYHQ>
> .
>

matt335672 commented:
That's a good point.

Yes, this is for VNC only. It brings resize on connect to VNC, but there's no reason why resize logic couldn't be used to implement #448 for VNC, provided the particular Xvnc server supports it.

Nexarian commented:
I really want to see this happen. How can I help?

On Thu, May 21, 2020 at 8:25 AM matt335672 <notifications@github.com> wrote:

> That's a good point.
>
> Yes, this is for VNC only. It brings resize on connect to VNC, but there's
> no reason why resize logic couldn't be used to implement #448
> <https://github.com/neutrinolabs/xrdp/issues/448> for VNC, provided the
> particular Xvnc server supports it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/neutrinolabs/xrdp/pull/1343#issuecomment-632057358>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UBHL2HYDOJXBVBHWMFCDRSUMULANCNFSM4HJ5BYHQ>
> .
>

matt335672 commented:
I think what I said was open to some misinterpretation - sorry about that.

What I meant was when #448 gets merged, getting the support into VNC too should be straightforward as all the functionality will be in the VNC library.

I can't comment on #448 - it somewhat predates my involvement. I can only suggest you ask Jay for an update on that thread. I'm happy to work on the VNC part, but I don't think that's at all critical.

metalefty commented:
At first glance, it is working quite well.

metalefty commented on vnc/vnc.c:
```diff
@@ -239,7 +273,7 @@ lib_process_channel_data(struct vnc *v, int chanid, int flags, int size,
                 // Send the RFB message type (CLIENT_CUT_TEXT) to the vnc server.
                 init_stream(out_s, clip_bytes + 1 + 3 + 4 + 16);
-                out_uint8(out_s, 6);   // RFB msg type:  CLIENT_CUT_TEXT
+                out_uint8(out_s, C2S_CLIENT_CUT_TEXT);
```
Thank you for this.

metalefty commented:
@matt335672 

Something going wrong when `-AcceptSetDesktopSize=0` (TigerVNC 1.7.0).

Regarding [community maintained RFB specification](https://github.com/rfbproto/rfbproto/blob/a2899832d9e6d4275819678bf0339af94e570e03/rfbproto.rst#screen-model), the VNC server may deny SetDesktopSize request.

```
[Xvnc]
param=Xvnc
param=-bs
param=-nolisten
param=tcp
param=-localhost
param=-dpi
param=96
param=-AcceptSetDesktopSize=0
```

```
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: connecting to sesman ip 127.0.0.1 port 3350
[20200522-13:16:06] [INFO ] xrdp_wm_log_msg: sesman connect ok
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: sending login info to session manager, please wait...
[20200522-13:16:06] [DEBUG] return value from xrdp_mm_connect 0
[20200522-13:16:06] [INFO ] xrdp_wm_log_msg: login successful for display 10
[20200522-13:16:06] [DEBUG] Layout from client_info (geom=1364x768 #screens=1) : 0:(1364x768+0+0)
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC started connecting
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC connecting to 127.0.0.1 5910
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC tcp connected
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC security level is 2 (1 = none, 2 = standard)
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC password ok
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC sending share flag
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC receiving server init
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC receiving pixel format
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC receiving name length
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC receiving name
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC sending pixel format
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC sending cursor
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: VNC connection complete, connected ok
[20200522-13:16:06] [DEBUG] xrdp_wm_log_msg: connected ok
[20200522-13:16:06] [DEBUG] xrdp_mm_connect_chansrv: chansrv connect successful
[20200522-13:16:06] [DEBUG] Closed socket 26 (AF_INET 127.0.0.1:41098)
[20200522-13:16:06] [DEBUG] Skipping ENC_CURSOR encoding
[20200522-13:16:06] [DEBUG] VNC matched ExtendedDesktopSize rectangle x=0, y=0 geom=1920x1080
[20200522-13:16:06] [DEBUG] X server supports resizing
[20200522-13:16:06] [DEBUG] Layout from Xvnc (geom=1920x1080 #screens=1) : 1804289383:(1920x1080+0+0)
[20200522-13:16:06] [DEBUG] VNC setting screen id to 1804289383 from server
[20200522-13:16:06] [DEBUG] Sending SetDesktopSize to change server layout
[20200522-13:16:06] [DEBUG] VNC Sending SetDesktopSize
[20200522-13:16:06] [DEBUG] Skipping ENC_RAW encoding
```

matt335672 commented:
Well this is interesting.

I agree with your reading of the spec, and I've managed to reproduce your results.

What seems to be happening is that when `AcceptDesktopSize=0` is set, the tigervnc server is still responding to our initial `SetEncodings` message with an `ExtendedDesktopSize` rectangle which advertises support for `SetDesktopSize`.

However, when we send `SetDesktopSize`, the server is dropping it. The relevant code for tigervnc is [here](https://github.com/TigerVNC/tigervnc/blob/6172cc01dd7d1493064571b341c978e2df7725d7/common/rfb/VNCSConnectionST.cxx#L618), with the message being dropped at line 624.

I think this is in violation of the spec in section 7.4.10. We should get an `ExtendedDesktopSize` rectangle with x set to 1 (error) and y set to 1 probably (Resize is administratively prohibited).

I'll raise an issue with tigervnc on this.

In the meantime, what do you think we should do? I can probably code around it, but it could be a bit messy.

metalefty commented on xrdp/xrdp.ini.in:
```diff
@@ -196,6 +196,10 @@ ip=127.0.0.1
 #xserverbpp=24
 #delay_ms=2000
+# Disable requested encodings to support buggy VNC servers
+# (1 = ExtendedDesktopSize)
```
In config files, we use semicolons to comment out descriptions or supplemental information.

metalefty replied:
This custom is started in #488.

metalefty commented on xrdp/xrdp.ini.in:
```diff
@@ -196,6 +196,10 @@ ip=127.0.0.1
 #delay_ms=2000
+# Disable requested encodings to support buggy VNC servers
+# (1 = ExtendedDesktopSize)
+#disabled_encodings_mask=0
```
This is fine because just removing `#` is still valid as configuration parameter.

metalefty commented:
Thank you for the investigation. 

I think we don't need to add a workaround if the code becomes messy.

As far as I observe, few people like the current behaviour but many people want to reconnect the existing VNC sessions. So I believe few people use `AcceptDesktopSize=0`. It is not worth code around than effort.

matt335672 commented:
Comments in xrdp.ini.in fixed.

I'll carry on looking at the tigervnc situation - I'd still like to check behaviour if a resize request is denied. I'll report back when I know more.

matt335672 commented:
I've built a tigervnc server [here](https://github.com/matt335672/tigervnc/tree/ack-set-desktop-size) which bounces SetDesktopSize requests with a denial if `AcceptSetDesktopSize=0` is set.

I've issued a PR for this, so hopefully I'll get some constructive feedback from the tigervnc project.

This has let me to make a few minor corrections which are committed above.

I've still got an unexplained problem in that the current latest remmina from the PPA (version 1.4.5+ppa202005242201.r36) on Bionic is occasionally closing on a resize with the following message:-

```
[ERROR][com.freerdp.core.capabilities] - unexpected MCS channel id 03ee received
```

It doesn't happen with any other clients. With remmina it doesn't happen all the time, so there's either uninitialised data or a race involved. In any case I think I'll need to bottom it out in case the problem is on our side.

metalefty commented:
Thanks, the TigerVNC team was so constructive when I sent a PR before.

Regarding Remmina issue, is it happen with FreeRDP?

matt335672 commented:
Yes - I've reproduced it with FreeRDP.

That should make debugging a bit easier.

matt335672 commented:
FreeRDP problem has been identified.

I've raised a separate PR #1595 to discuss, as it appears to be a protocol violation on our part (if I'm reading the spec right).

matt335672 commented:
For info, PR TigerVNC/tigervnc#1036 to respond to the SetDesktopSize request with a denial if `-AcceptSetDesktopSize=0` has been accepted.

metalefty commented:
Yup, I subscribed the PR and saw it was merged. Thank you very much for the fix.
Then, how much did you perform tests with non-resizable VNC server?

Some log messages should be promoted to INFO level.

metalefty commented on vnc/vnc.c:
```diff
@@ -586,6 +992,447 @@ make_color(int r, int g, int b, int bpp)
+    {
+        if (layout.s != NULL)
+        {
+            log_message(LOG_LEVEL_DEBUG, "X server supports resizing");
```
Doesn't "VNC server" sounds better?

metalefty commented on vnc/vnc.c:
```diff
@@ -586,6 +992,447 @@ make_color(int r, int g, int b, int bpp)
+        }
+        else
+        {
+            log_message(LOG_LEVEL_DEBUG, "X server does not support resizing");
```
Ditto.

matt335672 commented:
I've performed non-resizeable tests so far with:-
- x11vnc on the console
- A patched TigerVNC server with `-AcceptSetDesktopSize=0`
- the new setting `disabled_encodings_mask=1` in xrdp.ini

I'll try also with tight VNC and real VNC on Ubuntu 18.04. I don't know whether they support resizing or not, but I'll need to understand the behaviour before this goes live so we can support it.

metalefty commented on vnc/vnc.c:
```diff
@@ -255,7 +305,364 @@ lib_process_channel_data(struct vnc *v, int chanid, int flags, int size,
+ */
+static void
+log_debug_screen_layout(const char *source,
+                        const struct vnc_screen_layout *layout)
```
I would like to promote final resize result to INFO level. What do you think?

metalefty replied:
So, it would be nice to make this function to support multiple log level.

metalefty commented:
Thanks, that's nice. TigerVNC is the first implementation that supports resizing and AFAIK it is still the only implementation so far.

The fact the name of TigerVNC developer Pierre Ossman can be seen in RFC6143 shows that resizing is introduced to VNC by him.
https://tools.ietf.org/html/rfc6143#section-8.4.2

matt335672 commented:
Happy with all those comments above.

bolkedebruin commented:
@matt335672 awesome work! The code really got an upgrade. I was wondering if "online resize" is an option? I.e. The client resizes it's window, but without a reconnect, and the server adjusts it's viewport accordingly (xrandr).

metalefty commented:
@bolkedebruin It's out of scope of this PR. It would be done in #448 someday later. Let us focus on the VNC resizing itself.

metalefty commented on vnc/vnc.c:
```diff
@@ -646,8 +646,8 @@ resize_client_from_layout(struct vnc *v,
-            log_debug_screen_layout("OldLayout", layout);
-            log_debug_screen_layout("NewLayout", layout);
+            log_screen_layout(LOG_LEVEL_INFO, "OldLayout", layout);
+            log_screen_layout(LOG_LEVEL_INFO, "NewLayout", layout);
```
I think this should be LOG_LEVEL_ERROR since this block is for an error, right?

metalefty replied:
Always OldLayout == NewLayout. Either one should be client_layout?

matt335672 replied:
Good spot - thanks

metalefty commented on vnc/vnc.c:
```diff
@@ -255,7 +305,364 @@ lib_process_channel_data(struct vnc *v, int chanid, int flags, int size,
 }
 
-/******************************************************************************/
+/**************************************************************************//**
```
Is the last `//**` intended?

matt335672 replied:
It's a doxygen-compatible comment - see paragraph 4 of [https://www.doxygen.nl/manual/docblocks.html](https://www.doxygen.nl/manual/docblocks.html).

metalefty replied:
Understood. Doxygen is a nice tool to generate documents. IMO it's okay to mix the original style and Doxygen style.

metalefty commented on vnc/vnc.c:
```diff
@@ -586,6 +992,447 @@ make_color(int r, int g, int b, int bpp)
+            else
+            {
+                log_message(LOG_LEVEL_DEBUG, "Sending SetDesktopSize "
+                            "to change server layout");
```
This could cause redundant  logs. Just below `send_set_desktop_size()` writes almost the same log.

```
[20200601-12:58:03] [DEBUG] Sending SetDesktopSize to change server layout
[20200601-12:58:03] [DEBUG] VNC Sending SetDesktopSize
```

For consistency with the opposite condition, I'd like to have the message like "Server layout differs from the client layout, changing server layout"  here.

metalefty replied:
Also, isn't here the place to log OldLayout (current Xvnc layout)?

matt335672 replied:
Agree with both - OldLayout is now LOG_LEVEL_INFO

metalefty commented on vnc/vnc.c:
```diff
@@ -586,6 +992,447 @@ make_color(int r, int g, int b, int bpp)
+        {
+            if (response_code == 0)
+            {
+                log_message(LOG_LEVEL_DEBUG, "VNC server successfully resized");
```
I think here's the place to log NewLayout at INFO level because resizing has been finished here, right?

matt335672 replied:
Agreed

metalefty commented:
Wow, switching between multi screen and single screen (and plus resizing) works perfectly for me. Awesome.

matt335672 commented:
That's a good set of review comments - thanks very much.

I've incorporated those, and also added in the error strings for failed resizes. So now you get errors like the following:-

```
[20200602-16:41:00] [ERROR] VNC server resize failed - error code 1 [Resize is administratively prohibited]
```

As regards the Doxygen header comment, I'm thinking in the longer term Doxygen might be a good place to document some of the interfaces in the code, particularly the libraries. What do you think?

I've also done some testing with both tightvnc 1.3.10 and realVNC 4.1.1 on Ubuntu 18.04.

For both I created a single window of size 640x480, then re-connected at 800x600. With both servers I got a message like this:-

```
[20200602-17:27:34] [DEBUG] VNC server does not support resizing
[20200602-17:27:34] [DEBUG] Resizing client to server 640x480
```

which is as expected. Neither server responds to the ExtendedDesktopSize pseudo-encoding.

I had to set the Windows client to 24-bit depth rather than 32 to get tightvnc to work.

I've still got some more testing to do around colour depths - I'll get on to that tomorrow.

metalefty commented on vnc/vnc.c:
```diff
@@ -1408,12 +1437,14 @@ lib_framebuffer_waiting_for_resize_confirm(struct vnc *v)
-                            response_code);
+                            "VNC server resize failed - error code %d [%s]",
+                            response_code,
+                            get_eds_status_msg(response_code));
```
It became much informative, thanks!

metalefty commented:
> As regards the Doxygen header comment, I'm thinking in the longer term Doxygen might be a good place to document some of the interfaces in the code, particularly the libraries. What do you think?

I totally agree. Good documentation will invite more first-time contributors.

Regarding TightVNC 32bpp issue, supporting multiple VNC implementations is definitely a good thing but I don't think we need to care about relatively tiny issues such as 32bpp if the fix is not easy. Open-source version of TightVNC has been inactive for a decade and it is the biggest reason that TigerVNC is forked from TightVNC in 2009.

Overall looks much better to me. We can make further fixes later. When you finished some more tiny fixes, let's merge.

matt335672 commented:
I've done some more testing this morning, and I've not hit any more issues.

@bolkedebruin - thanks for following this. Have you got anything to add before we merge, in a day or so?

bolkedebruin commented:
We can test it on ca 200+ users on Monday if you are willing to wait. Can also file bugs afterwards

metalefty commented:
@bolkedebruin If you could test it, it would be nice to test devel branch after we merge this. If you find bugs, raise a new issue.

matt335672 commented:
This branch is currently baselined off 0.9.12.

[v0.9.13](../releases/tag/v0.9.13) is bugfixes, and a scan of the commit log since then shows these PRs. I've checked this list once, but I wouldn't rely on it without checking yourself:-
- #1529 - [bugfix] Fix Xorg path for Arch Linux
- #1522 - [bugfix] workaround odd shift behaviour
- #1560 - [feature] Support Slackware with pam
- #1557 - [security] Check for overflow before g_malloc() calls
- #1549 - [security] Address possible memory out-of-bounds accesses on connect requests
- #1520 - [bugfix] Small fixes found by static analyzers
- #1593 - [bugfix] Ignore TS_MULTIFRAGMENTUPDATE_CAPABILITYSET from client if fp disabled
- #1595 - [bugfix] Prevent chansrv input channels being scanned during a server reset

They're all reasonably tiny. The only coupling with this PR that I'm aware of is that #1595 is required if you're using xfreerdp or Remmina clients, and you're using a non-resizing VNC server.

@bolkedebruin - we're really grateful of an offer of testing like this - thank you. Clearly we'd prefer it if you tested devel after a merge. Whether you'd be prepared to do that or not has to be your decision - you know your workflows and we don't.

I'm going to be a bit strapped for time tomorrow, so I'll do the merge into devel today. I'll leave the branch active in my repo for the time being, so you can pick whatever suits you for your testing. The branch will be squashed, but will still be based off v0.9.12.

One other minor point - if you've been using my original PR with SCP_SESSION_TYPE_RESIZEABLE_VNC, you'll need to remove the `code=1` setting from `xrdp.ini`. I've mentioned this before, but it's worth re-iterating.

bolkedebruin commented:
@matt335672 we are fine using the devel branch after merging and report bugs from there. We can revert our deployments quite easily. Note that we will just test expected behavior (happy flow) as obviously we want to keep disruption to the users low.

Our setup looks like "browser -> guacamole -> xrdp -> tigervnc -> X -> container -> kubernetes"

matt335672 commented:
Well that makes my life easier then - I've rebased the branch to devel before the merge, as this will make the history cleaner.

For the record, as well as the squash I've updated manpages:-
-  `xrdp.ini(5)` now documents the `disabled_encodings_mask` setting
- `sesman.ini(5)` documents the changed behaviour to the `Policy` setting.

I've also added a [wiki page](https://github.com/neutrinolabs/xrdp/wiki/Xvnc-backend---Multi-monitor-and-resize-support) which details the new behaviour.

metalefty commented on docs/man/sesman.ini.5.in:
```diff
@@ -148,12 +148,12 @@ This works only with xorgxrdp session. Moreover, xorgxrdp must be v0.2.9 or late
-for the combination <User,BitPerPixel> when using Xrdp, and
-for the combination <User,BitPerPixel,DisplaySize> when using Xvnc.
-This behavior can be changed by setting session policy to:
+Session allocation policy. Used to decide when to allocate a
```
Thank you for caring about docs!

bolkedebruin commented:
All seems fine with testing. We had some 70 concurrent users and we didnt receive any reports that pointed to issues.

matt335672 commented:
@bolkedebruin - thanks ever so much.

There's really no substitute for large scale testing and it was really good of you to fit us in to your schedule like this.
